In [2]:
import numpy as np

In [3]:
import pandas as pd
from sklearn.preprocessing import StandardScaler  #used to standardize data to a common range
from sklearn.model_selection import train_test_split

In [4]:
from sklearn import svm #classifier, will get fed clean data to predict outcome

In [5]:
from sklearn.metrics import accuracy_score


DATA COLLECTION AND EDA
0-->non diabetic
1-->diabetic

In [8]:
#loading dataset to pandas dataframe
diab = pd.read_csv("/content/diabetes (1).csv")

In [10]:
diab.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

THEREFORE WE NEED TO DEVELOP A SYSTEM THAT WILL CLASSIFY THESE ATTRIBUTES INTO A SINGLE OUTPUT 1(DIABETIC) OR 0 (NON DIABETIC)

In [9]:
diab.shape

(768, 9)

In [11]:
#statistical measures
diab.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  768.000000                768.000000  768.000000  768.000000  
mean    31.992578                  0.471876   33.240885    0.348958  
std      7.884160                  0.331329   11.760232    0.476951  
min      0.000000                  0.078000   21.000000    0.000000  
25%     27.300000                  0.243750   24.000000    0.000000  
50%     32.000000                  0.372500   29.000000    0.000000  
75%     36.600000                  0.626250   41.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

In [12]:
diab['Outcome'].value_counts()

Outcome
0    500
1    268
Name: count, dtype: int64

500 non diabetic and 268 diabetic values in this dataset

In [13]:
diab.groupby('Outcome').mean()
# this will give mean values of all attributes for the OUTCOME 0 & 1
# thus helping us recognize patterns in values of glucose, BP etc for non diabetic and diabetic people

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
Outcome                                                                      
0           3.298000  109.980000      68.184000      19.664000   68.792000   
1           4.865672  141.257463      70.824627      22.164179  100.335821   

               BMI  DiabetesPedigreeFunction        Age  
Outcome                                                  
0        30.304200                  0.429734  31.190000  
1        35.142537                  0.550500  37.067164

In [22]:
X = diab.drop(columns = 'Outcome', axis =1)
Y = diab['Outcome']
#separating outcome column in X and storing in Y

In [23]:
X

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6      148             72             35        0  33.6   
1              1       85             66             29        0  26.6   
2              8      183             64              0        0  23.3   
3              1       89             66             23       94  28.1   
4              0      137             40             35      168  43.1   
..           ...      ...            ...            ...      ...   ...   
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  
0                       0.627   50  
1                       0.351   31  
2                       0.672   32  
3                       0.167   21  
4                       2.288   33  
..                        ...  ...  
763                     0.171   63  
764                     0.340   27  
765                     0.245   30  
766                     0.349   47  
767                     0.315   23  

[768 rows x 8 columns]

In [24]:
Y

0      1
1      0
2      1
3      0
4      1
      ..
763    0
764    0
765    0
766    1
767    0
Name: Outcome, Length: 768, dtype: int64

STANDARDIZING DATA: since the range of values of all attributes is different, it is necessary to standardize values for our machine learning model

In [26]:
scaler = StandardScaler()
#running 1 instance of standardscaler

In [29]:
scaler.fit(X)
#fitting data

StandardScaler()

In [32]:
Standardized_data = scaler.transform(X)
#transform data

In [33]:
Standardized_data

array([[ 0.63994726,  0.84832379,  0.14964075, ...,  0.20401277,
         0.46849198,  1.4259954 ],
       [-0.84488505, -1.12339636, -0.16054575, ..., -0.68442195,
        -0.36506078, -0.19067191],
       [ 1.23388019,  1.94372388, -0.26394125, ..., -1.10325546,
         0.60439732, -0.10558415],
       ...,
       [ 0.3429808 ,  0.00330087,  0.14964075, ..., -0.73518964,
        -0.68519336, -0.27575966],
       [-0.84488505,  0.1597866 , -0.47073225, ..., -0.24020459,
        -0.37110101,  1.17073215],
       [-0.84488505, -0.8730192 ,  0.04624525, ..., -0.20212881,
        -0.47378505, -0.87137393]])

this will help our model make better predictions as all the values are now in a similar range

In [34]:
X = Standardized_data
Y = diab['Outcome']

In [35]:
X

array([[ 0.63994726,  0.84832379,  0.14964075, ...,  0.20401277,
         0.46849198,  1.4259954 ],
       [-0.84488505, -1.12339636, -0.16054575, ..., -0.68442195,
        -0.36506078, -0.19067191],
       [ 1.23388019,  1.94372388, -0.26394125, ..., -1.10325546,
         0.60439732, -0.10558415],
       ...,
       [ 0.3429808 ,  0.00330087,  0.14964075, ..., -0.73518964,
        -0.68519336, -0.27575966],
       [-0.84488505,  0.1597866 , -0.47073225, ..., -0.24020459,
        -0.37110101,  1.17073215],
       [-0.84488505, -0.8730192 ,  0.04624525, ..., -0.20212881,
        -0.47378505, -0.87137393]])

In [36]:
Y

0      1
1      0
2      1
3      0
4      1
      ..
763    0
764    0
765    0
766    1
767    0
Name: Outcome, Length: 768, dtype: int64

TRAIN TEST SPLIT

In [50]:
X_train, X_test, Y_train, Y_test = train_test_split(X,Y,test_size=0.2, stratify = Y, random_state=2)
#test size = 0.2 means 20% will be used for testing and 80% for training
#stratify = ensures that the split has the same proportion of the target outcome as the original dataset
# => if for example 80% are diabetic and 20% are non diabetic in the original dataset, then the same proportion will be ensured in the splits
# random state is used for data shuffling to improve accuracy

In [40]:
X.shape

(768, 8)

In [41]:
X_train.shape

(614, 8)

In [42]:
X_test.shape

(154, 8)

In [68]:
Y.shape


(768,)

In [66]:
Y_train.shape

(614,)

In [67]:
Y_test.shape

(154,)

this denotes the proportion used for training and testing from the original

In [44]:
classifier=svm.SVC(kernel='linear')
# linear kernel works well when the classes in the dataset are linearly separable, meaning they can be separated by a straight line/hyperplane.
# The decision boundary created by a linear kernel is a straight line (or hyperplane) that best separates the classes based on the given feature vectors.

In [45]:
#training the SVM classifier
classifier.fit(X_train, Y_train)


SVC(kernel='linear')

Model Evaluation

In [46]:
#ACCURACY SCORE
X_train_prediction = classifier.predict(X_train)
#this will predict the labels (output 0 or 1) for all X_train and store in X_train_prediction
X_train_prediction

array([0, 0, 1, 0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0,
       0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1,
       0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 1, 0, 1,
       1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 1, 0,
       1, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0,
       1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0,
       1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1,
       0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1,
       0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1,

In [47]:
training_data_accuracy= accuracy_score(X_train_prediction, Y_train)
#this compares these above predicted values of X_train_prediction with original dataset values ie Y_train

In [48]:
training_data_accuracy

0.7866449511400652

this means, out of every 100 predictions, 78 are correct which is a decent score

78% training data accuracy

now same process for test data

In [51]:
classifier.fit(X_test, Y_test)

SVC(kernel='linear')

In [55]:
X_test_prediction = classifier.predict(X_test)
X_test_prediction
##this will predict the labels (output 0 or 1) for all X_test and store in X_test_prediction

array([0, 0, 1, 1, 0, 0, 0, 1, 0, 1, 1, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1,
       1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 1, 1, 0,
       0, 1, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0,
       1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 1,
       1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1])

In [56]:
test_data_accuracy= accuracy_score(X_test_prediction, Y_test)
##this compares these above predicted values of X_test_prediction with original dataset values ie Y_test

In [57]:
test_data_accuracy

0.7532467532467533

75% testing data accuracy

MAKING A PREDICTIVE SYSTEM


In [63]:
input_data = (4,110,92,0,0,37.6,0.191,30)

#NOTE:TARGET VARIABLE OUTPUT IS NOT SELECTED AS IT IS TO BE PREDICTED

#changing input data to numpy array
input_data_as_numpy_array=np.asarray(input_data)
#np.asarray function is used for conversion
input_data_as_numpy_array

array([  4.   , 110.   ,  92.   ,   0.   ,   0.   ,  37.6  ,   0.191,
        30.   ])

In [60]:
input_data_reshaped = input_data_as_numpy_array.reshape(1,-1)
#reshaping can be useful when you need to convert a single row of data into a format that some machine learning models expect
#reshape(1, -1) transforms arr_1d into a 2-dimensional array (arr_2d) with one row (1) and enough columns (-1) to accommodate all elements of arr_1d.
input_data_reshaped

array([[  4.   , 110.   ,  92.   ,   0.   ,   0.   ,  37.6  ,   0.191,
         30.   ]])

In [61]:
#HOWEVER WE NEED STANDARDIZED DATA AS INPUT
standard_data= scaler.transform(input_data_reshaped)
standard_data

/usr/local/lib/python3.10/dist-packages/sklearn/base.py:439: UserWarning: X does not have valid feature names, but StandardScaler was fitted with feature names
  warnings.warn(


array([[ 0.04601433, -0.34096773,  1.18359575, -1.28821221, -0.69289057,
         0.71168975, -0.84827977, -0.27575966]])

In [62]:
prediction = classifier.predict(standard_data)
prediction

#classifier.predict function is used

array([0])

THEREFORE THE MODEL HAS PREDICTED CORRECTLY

In [71]:
if (prediction[0]==0):
  print(" NON DIABETIC :)")
else:
  print("NON DIABETIC :(")

 NON DIABETIC :)


TRY FOR DIABETIC INSTANCE BY FEEDING THOSE VALUES AS INPUT DATA

In [5]:
from flask import Flask, request, jsonify
import joblib

# Assuming 'model' is your trained machine learning model
joblib.dump(model, 'my_model.joblib')

NameError: name 'model' is not defined